packages/dds/tree/src/forest/cursor.ts

```diff
@@ -4,125 +4,236 @@
- * Should either be documented here, or each producer should document them
- * (and likely via returning a sub-interface with documentation on the subject).
- *
- * TODO: Needs a way to efficiently clone Cursor so patterns like lazy tree reification can be implemented efficiently.
- *
- * TODO: add optional fast path APIs for more efficient handling when supported by underlying format and reader.
- * Leverage "chunks" and "shape" for this, and skip to next chunk with seek (chunk length).
- * Default chunks of size 1, and "node" shape?
- */
-export interface ITreeCursor<TResult = TreeNavigationResult> {
     /**
-     * Select the child located at the given key and index.
+     * @returns the number of immediate children in the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
      */
-    down(key: FieldKey, index: number): TResult;
+    getCurrentFieldLength(): number;
 
     /**
-     * Moves `offset` entries in the field.
+     * Sets current node to the node at the provided `index` of the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     * Sets mode to `Nodes`.
      */
-    seek(offset: number): TResult;
+    enterChildNode(index: number): void;
+
+    // ********** APIs for when mode = Nodes ********** //
 
     /**
-     * Select the parent of the currently selected node.
+     * @returns a path to the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     * Assumes this cursor has a root node where its field keys are actually detached sequences.
+     * If the cursor is not rooted at such a node,
+     * calling this function is invalid, and the returned UpPath (if any) may not be meaningful.
+     * This requirement exists because {@link UpPath}s are absolute paths
+     * and thus must be rooted in a detached sequence.
+     * TODO: consider adding an optional base path to append to remove/clarify this restriction.
      */
-    up(): TResult;
+    getPath(): UpPath | undefined;
+
+    /**
+     * Index (within its parent field) of the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentIndexInField: number;
+
+    /**
+     * Index (within its parent field) of the first node in the current chunk.
+     * Always less than or equal to `currentIndexInField`.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentChunkStart: number;
+
+    /**
+     * Length of current chunk.
+     * Since an entire chunk always has the same `pending` value,
+     * can be used to help skip over all of a pending chunk at once.
+     *
+     * TODO:
+     * Add optional APIs to access underlying chunks so readers can
+     * accelerate processing of chunk formats they understand.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentChunkLength: number;
+
+    // ********** APIs for when mode = Nodes and not pending ********** //
```

Comment:
The APIs are not currently organized to match this. Either reorganize them or remove this.
```suggestion
    // ********** APIs for when mode = Nodes and not pending ********** //
```

Reply:
I reorganized them.